connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java

```diff
@@ -0,0 +1,581 @@
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        MultiFieldPaths that = (MultiFieldPaths) o;
+        return Objects.equals(trie, that.trie);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(trie);
+    }
+
+    @Override
+    public String toString() {
+        return "FieldPaths(trie = " + trie + ")";
+    }
+
+    // Invariants:
+    // - Trie values contain either a nested trie or a field path when it is a leaf.
+    // - A trie flattens overlapping paths (e.g. `foo` and `foo.bar` in V2, only `foo.bar` would be kept)
+    static class Trie {
+        TrieNode root;
+
+        Trie() {
+            root = new TrieNode();
+        }
+
+
+        public void insert(SingleFieldPath path) {
+            TrieNode current = root;
+
+            for (String step : path.stepsWithoutLast()) {
+                if (!current.contains(step)) {
+                    current.addStep(step);
+                }
+
+                current = current.get(step);
+            }
+
+            final String step = path.lastStep();
+            if (!current.contains(step)) {
+                current.addLeaf(step, path);
+            }
+        }
+
+        public boolean isEmpty() {
+            return root.isEmpty();
+        }
+
+        public Optional<TrieNode> find(String step) {
+            return root.find(step);
+        }
+
+        public int size() {
+            if (root.isEmpty()) return 0;
+            return root.size();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Trie trie = (Trie) o;
+            return Objects.equals(root, trie.root);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(root);
+        }
+
+        @Override
+        public String toString() {
+            return "Trie(" +
+                "root = " + root +
+                ')';
+        }
+    }
+
+    static class TrieNode {
+        Map<String, TrieNode> steps = new HashMap<>();
+        SingleFieldPath path;
+
+        TrieNode() {
+        }
+
+        private TrieNode(SingleFieldPath path) {
+            this.path = path;
+        }
+
+        public boolean contains(String step) {
+            return steps.containsKey(step);
+        }
+
+        public void addStep(String step) {
+            if (path != null) path = null;
+            steps.put(step, new TrieNode());
+        }
+
+        public void addLeaf(String step, SingleFieldPath path) {
+            if (this.path != null) this.path = null;
+            steps.put(step, new TrieNode(path));
+        }
+
+        public TrieNode get(String step) {
+            return steps.get(step);
+        }
+
+        public Optional<TrieNode> find(String step) {
+            return Optional.ofNullable(steps.get(step));
+        }
```

Comment:
Aren't these two methods essentially identical? `get` will return `null` if the step isn't found, `find` will return an empty `Optional`.

IMO we should either get rid of `find` entirely and just do standard null checks for return values from it when calling it, or alter `get` to throw an exception if the step _must_ be present but for some reason isn't.

Reply:
Agree, I agree the Optional variants do not make sense as not exposed to users, but used only by MultiFieldPath. Applying suggestion